File: algo_expert/sorting_algorithms/merge_sort.js



// TC O(n * logn) SC O(n)
function mergeSort(arr){

    if (arr.length === 1){
        return arr;
    }

    let midPoint = Math.floor(arr.length / 2)
    let left = arr.slice(0, midPoint)
    let right = arr.slice(midPoint)

    console.log(`Left side ${left}`);
    console.log(`Right side ${right}`);

    return merge(
        mergeSort(left),
        mergeSort(right)
    )
}


function merge(left, right){
    let outputArr = []
    let leftIdx = 0;
    let rightIdx = 0;
    while (leftIdx < left.length  && rightIdx < right.length){
        if (left[leftIdx] <= right[rightIdx]){
            outputArr.push(left[leftIdx]);
            leftIdx++;
        } else{
            outputArr.push(right[rightIdx]);
            rightIdx++;
        }
    }
    
    let mergedArr = outputArr.concat(left.slice(leftIdx)).concat(right.slice(rightIdx))
    console.log(`merge Right side ${right} and Left side ${left}`);
    console.log(`mergedArr ${mergedArr}`);
    return mergedArr
}










let unsortedArray = [7, 3, 1, 4 ,2];
console.log(`Unsorted array ${unsortedArray}`)
console.log(`Sorted array mergeSort ${mergeSort(unsortedArray)}`)

File: algo_expert/sorting_algorithms/merge_sort.py
from typing import List, Union


# O(n *logn) SC O(n)
def mergeSort(arr: List) -> List:
    
    if len(arr) == 1:
        return arr
    
    mid_idx = len(arr) // 2
    left_part = arr[:mid_idx]
    right_part = arr[mid_idx:]

    return merge(
        mergeSort(left_part),
        mergeSort(right_part)
    )


def merge(left, right):

    merged_arr = []
    left_idx = 0
    right_idx = 0
    while left_idx < len(left) and right_idx < len(right):
        left_value = left[left_idx]
        right_value = right[right_idx]
        if left_value < right_value:
            merged_arr.append(left_value)
            left_idx += 1
        else:
            merged_arr.append(right_value)
            right_idx += 1            

    return merged_arr + left[left_idx:] + right[right_idx:]


input_arr = [4, 54, 1, 0, -4, 9]
print(f"Input array goes as follows {input_arr}")
print(f"Sorted array {mergeSort(input_arr)}")


input_arr = [2, 1, 5]
print(f"\n\n Input array goes as follows {input_arr}")
print(f"Sorted array {mergeSort(input_arr)}")